chain/src/main/scala/org/bitcoins/chain/blockchain/BlockchainUpdate.scala

```diff
@@ -14,22 +15,59 @@ import org.bitcoins.core.protocol.blockchain.BlockHeader
   * because of [[org.bitcoins.chain.validation.TipUpdateResult.BadPOW BadPOW]] or a
   * [[org.bitcoins.chain.validation.TipUpdateResult.BadNonce BadNonce]] etc
   */
-sealed abstract class BlockchainUpdate
+sealed abstract class BlockchainUpdate {
+  def successfulHeaders: Vector[BlockHeaderDb]
+
+  def blockchain: Blockchain
+
+  def withSuccessfulHeaders(headers: Vector[BlockHeaderDb]): BlockchainUpdate =
+    this match {
+      case s: Successful =>
+        s.copy(successfulHeaders = headers)
+      case f: Failed =>
+        f.copy(successfulHeaders = headers)
+    }
+}
 
 object BlockchainUpdate {
 
   /** The key thing we receive here is [[org.bitcoins.chain.models.BlockHeaderDb BlockHeaderDb]]
     * with a height assigned to it this happens after
-    * calling [[org.bitcoins.chain.blockchain.ChainHandler.processHeader ChainHandler.processHeader]]
+    * calling [[org.bitcoins.chain.blockchain.ChainHandler.processHeaders ChainHandler.processHeaders]]
     */
-  case class Successful(blockchain: Blockchain, updatedHeader: BlockHeaderDb)
+  case class Successful(
+      blockchain: Blockchain,
+      successfulHeaders: Vector[BlockHeaderDb])
       extends BlockchainUpdate {
-    def height: Long = updatedHeader.height
+    if (successfulHeaders.nonEmpty) {
+      require(
+        blockchain.tip == successfulHeaders.head,
+        s"Tip did not equal last successful header, tip=${blockchain.tip.hashBE} lastSuccessfulHeader=${successfulHeaders.head.hashBE}"
+      )
+    }
+    def height: Long = blockchain.height
   }
 
+  /**
+    * Means we failed to update the given blockchain with _ALL_ given headers
+    * This means we could have had a partially successful update, with the headers/blockchain
+    * returned in this case class
+    */
   case class Failed(
       blockchain: Blockchain,
+      successfulHeaders: Vector[BlockHeaderDb],
       failedHeader: BlockHeader,
       tipUpdateFailure: TipUpdateResult.Failure)
-      extends BlockchainUpdate
+      extends BlockchainUpdate {
+    require(
+      !blockchain.contains(failedHeader),
+      s"Our blockchain should not contain the failed header=${failedHeader}")
+
+    if (successfulHeaders.nonEmpty) {
+      require(
+        successfulHeaders.head == blockchain.tip,
+        s"Our blockchain.tip should be the first successful header, got=${blockchain.tip} expected=${successfulHeaders.head}"
```

Comment:
Would it move to work these duplicated `require` statements into the abstract class, with a match on `this` to get the correct error message?